### PR TITLE
:construction_worker: feat(ci): add semantic releases

### PR DESCRIPTION
**Pull request by Tifloz**
# Pull request for Candy-doc

## Your checklist for this pull request

🚨 Please review the [guidelines for contributing](../../CONTRIBUTING.md) to
this repository.

- [x] Make sure you are requesting to **pull a feature/bugfix branch** (right
  side). Don't request your master!
- [x] Make sure you are making a pull request against the **main branch** (left
  side).
- [x] Check the commit's or even all commits' message styles matches our
  requested structure.
- [x] Check your code additions will fail neither code linting checks nor unit
  test.

## Description

We need to add a semantic release workflow to deliver a correctly tagged version each time we accept a PR

❤️ Thank you!

### CODE REVIEW

**Review by ecattez — CHANGES_REQUESTED** [review]
We have to define the keywords to release a major / minor / patch version

File: .github/workflows/publish_release.yml [hunk 1]
```diff
@@ -0,0 +1,37 @@
+name: Publish a new release
+on:
```
[the remark this answers is not in the document]
Comment (Tifloz) [reply .github/workflows/publish_release.yml:3]: Done

File: .releaserc.json [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+        "config": "conventional-changelog-gitmoji-config"
+      }
+    ],
+    [
+      "@semantic-release/git",
+      {
+        "assets": [
+          "CHANGELOG.md"
+        ],
+        "message": ":bookmark: chore(release): ${nextRelease.gitTag} [skip ci] \n\n${nextRelease.notes}"
```
Comment (ecattez) [inline .releaserc.json:38]: If "skip ci" is configurable, use `chore(release)` as the pattern to skip the ci.

File: .github/workflows/publish_release.yml [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+  workflow_dispatch:
+    inputs:
+      isMajor:
+        type: boolean
+        required: false
+        description: This is a new major version
+      message:
+        type: string
+        required: false
+        description: Commit message
```
Comment (ecattez) [inline .github/workflows/publish_release.yml:13]: Breaking changes message

File: .github/workflows/publish_release.yml [hunk 4]
```diff
@@ -0,0 +1,79 @@
+name: Publish a new release
+
+on:
+  workflow_dispatch:
+    inputs:
+      isMajor:
+        type: boolean
+        required: false
+        description: This is a new major version
+      message:
```
Comment (ecattez) [inline .github/workflows/publish_release.yml:10]: breaking_changes

File: .github/workflows/publish_release.yml [hunk 5]
```diff
@@ -0,0 +1,79 @@
+name: Publish a new release
+
+on:
+  workflow_dispatch:
+    inputs:
+      isMajor:
```
Comment (ecattez) [inline .github/workflows/publish_release.yml:6]: is_major

File: .github/workflows/publish_release.yml [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+        type: boolean
+        required: false
+        description: This is a new major version
```
Comment (ecattez) [inline .github/workflows/publish_release.yml:9]: Make a major version

File: .github/workflows/publish_release.yml [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+
+    - name: Set up JDK 11
+      uses: actions/setup-java@v3
+      with:
+        java-version: '11'
+        distribution: adopt
+        cache: maven
+
+    - name: Build with Maven
+      run: mvn --batch-mode -DskipTests package --update-snapshots verify
```
Comment (ecattez) [inline .github/workflows/publish_release.yml:31]: Make a composition action to use those jobs in release and in build. Check other jobs if there is other possible compositions

**Review by ecattez — CHANGES_REQUESTED** [review]

File: .github/workflows/publish_release.yml [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+      with:
+        extra_plugins: |
+          "semantic-release-config-gitmoji"
+          "semantic-release-maven --legacy-peer-deps"
+          "@conveyal/maven-semantic-release"
+        extends: semantic-release-config-gitmoji
+      env:
+        GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+        MVN_CENTRAL_LOGIN: ${{ secrets.MVN_CENTRAL_LOGIN }}
+        MVN_CENTRAL_PASS: ${{ secrets.MVN_CENTRAL_PASS }}
```
Comment (ecattez) [inline .github/workflows/publish_release.yml:75]: PASS -> PWD

File: pom.xml [hunk 9]
```diff
@@ -8,6 +8,76 @@
...
+            <email>sregnault@ippon.fr</email>
+            <organization>Ippon</organization>
+            <organizationUrl>https://fr.ippon.tech/</organizationUrl>
+        </developer>
+        <developer>
+            <name>Charles Tacquet</name>
+            <email>ctacquet@ippon.fr</email>
+            <organization>Ippon</organization>
+            <organizationUrl>https://fr.ippon.tech/</organizationUrl>
+        </developer>
```
Comment (ecattez) [inline pom.xml:60]: Add @aginesy @AVernholles and @Unknow46

File: pom.xml [hunk 10]
```diff
@@ -8,6 +8,76 @@
...
+    </developers>
+
+    <distributionManagement>
+        <snapshotRepository>
+            <id>ossrh</id>
+            <url>https://s01.oss.sonatype.org/content/repositories/snapshots</url>
```
Comment (ecattez) [inline pom.xml:66]: Check with @ctacquet if we succeed to pull image from maven central

**Review by ecattez — CHANGES_REQUESTED** [review]

File: pom.xml (hunk 10 above)
Comment (Tifloz) [reply pom.xml:66]: It works 🎉

**Comment by Tifloz** [comment]
@Michiels96  Can you explain how to setup secrets used in your workflow and how to populate them ?

File: pom.xml (hunk 9 above)
Comment (Michiels96) [reply pom.xml:60]: Done

File: README.md [hunk 11]
```diff
@@ -241,6 +242,19 @@ Explanations:
...
+
+### Fork Candy-doc
+
+The .github/workflows/publish-release.yml file uses these secrets :
+- GPG_PRIVATE_KEY
+- GPG_PASSPHRASE
+- MVN_CENTRAL_LOGIN
+- MVN_CENTRAL_PWD
+- GPG_KEY_NAME
+- GPG_PASSPHRASE
```
Comment (Tifloz) [inline README.md:254]:
```suggestion
```
Duplicated

**Review by Michiels96 — DISMISSED** [review]

File: README.md [hunk 12]
```diff
@@ -241,6 +242,18 @@ Explanations:
...
+### Fork Candy-doc
+
+The .github/workflows/publish-release.yml file uses these secrets :
+- GPG_PRIVATE_KEY
+- GPG_PASSPHRASE
+- MVN_CENTRAL_LOGIN
+- MVN_CENTRAL_PWD
+- GPG_KEY_NAME
+
+When you fork this project, add these secrets in your forked repository settings in order to publish your own version of Candy-doc on Maven Central.
```
Comment (ecattez) [inline README.md:255]: I don't think we want people to publish release of Candy-Doc on Maven Central. A message about "you will not able to publish release in a fork" is a better option in my opinion

**Review by ecattez — CHANGES_REQUESTED** [review]

File: README.md (hunk 12 above)
Comment (Tifloz) [reply README.md]: Done @ecattez

File: .github/workflows/publish_release.yml [hunk 13]
```diff
@@ -0,0 +1,77 @@
+name: Publish a new release
```
Comment (ecattez) [inline .github/workflows/publish_release.yml:1]: Two different workflows: one for manual releases, and an other for automatic releases

**Review by ecattez — CHANGES_REQUESTED** [review]

File: .github/workflows/publish_release.yml (hunk 13 above)
Comment (Tifloz) [reply .github/workflows/publish_release.yml:1]: Done

**Review by ecattez — APPROVED** [review]

**Review by ctacquet — APPROVED** [review]

**Pull request merged**